src/enumerable.cr

```diff
@@ -2292,7 +2298,11 @@ module Enumerable(T)
     # if the type is a union.
     def self.first
       {% if X.union? %}
-        {{X.union_types.first}}
+        {{
+          raise("Enumerable#sum/product() does support Union types. " +
+                "Instead, use Enumerable#sum/product(initial) with an " +
+                "initial value of the expected type of the sum/product call.")
```

Comment:
**suggestion:** Polish the error message:
```suggestion
          raise("`Enumerable#sum` and `#product` cannot determine the initial value from a union type. " +
                "Please pass an initial value of the intended type to the call.")
```

Reply:
After reading your fix, I realized a missing negation in the first sentence :)

Also, how about we update the error message as follows?

> raise("Enumerable#sum` and #product  do not support union types.  Instead, use Enumerable#sum/product(initial) with an initial value of the expected type of the sum/product call.")

Following is my reasoning for the above change.
1. The existing logic fails to identify an appropriate/safe type that can hold the final value of the call.
2. Using the identity is a way of determining the appropriate/safe type.  Since it is an implementation detail, I think we should surface such details in documentation while mentioning what (not how) is wrong and its fix in the error message.

Reply:
"do not support union types" isn't entirely accurate. They do support them. Just without automatically determining the return type, hence the need to specify explicitly.

Reply:
With this change, "sum()" and "product()" will not support union types (but "sum(initial)" and "product(initial)" will support union types).  So, do you mean that we should be precise in the error message and the doc by referring to the no-arg variant of the methods?

Reply:
Yes, we could differentiate between the different overloads by signature.

> `Enumerable#sum()` and `#product()` do not support union types.`

It's a bit of a subtle detail though. Hence my suggestion to write out the issue explicitly:

> `Enumerable#sum` and `#product` cannot determine the initial value from a union type.

I'm fine either way though.

Reply:
I have updated the PR such that 
1. error messages and doc refer to specific overloads and 
2. error messages focus only on the what is wrong and how to fix it.